### PR TITLE
Keyword display improvements

### LINKED ISSUE #15639: MethodError message misleading/wrong on 0.4, very hard to read on 0.5
Reproduce with
```julia
type FooCon
    x
end
type BarCon
    x
end
function addConstraint(c::FooCon)
    println("addCon(FooCon)")
end
function addConstraint(c::BarCon; uncset=nothing)
    println("addCon(BarCon,$uncset)")
    # Oh, BarCon was suppossed to be a FooCon, let me erroneously pass
    # on this kwarg which isn't supported
    addConstraint(FooCon(c.x), uncset=uncset)
end
addConstraint(BarCon(1))
```

On Julia 0.4.5:
```
ERROR: LoadError: MethodError: `addConstraint` has no method matching addConstraint(::FooCon)
Closest candidates are:
  addConstraint(::FooCon)
 in include at ./boot.jl:261
 in include_from_node1 at ./loading.jl:304
 in process_options at ./client.jl:280
 in _start at ./client.jl:378
```

Julia 0.5 (from January 31st)
```
ERROR: LoadError: MethodError: `#addConstraint` has no method matching #addConstraint(::Array{Any,1}, ::#addConstraint, ::FooCon)
Closest candidates are:
  #addConstraint(::Array{T,N}, ::#addConstraint, ::BarCon)
 [inlined code] from ./boot.jl:331
 in #addConstraint#1(##addConstraint#1, Void, Any, BarCon) at /Users/idunning/.julia/v0.4/JuMPeR/rep.jl:14
 in include(Core.#include, ASCIIString) at ./boot.jl:264
 in include_from_node1 at ./loading.jl:417
 in process_options at ./client.jl:244
 in _start at ./client.jl:300
```

### PR DESCRIPTION
**Pull request by dhoegh**
This PR implements several improvement to how keywords is presented to the user. The additions in each commit is listed bellow
- 1. adds display of variable keyword arguments such as `kwargs...`. This change causes the vararg keywords to be displayed in method list and other places.
- 2. adds the display of keywords to method completion 

```
julia> split(<tab>
split{T<:SubString{T<:AbstractString}}(str::T, splitter; limit, keep) at strings/util.jl:126
split(str::AbstractString) at strings/util.jl:151
split{T<:AbstractString}(str::T, splitter; limit, keep) at strings/util.jl:127
```
- 3. Make the generic error thrown by `kwerr` throw a MethodError. This allows for display of other method candidates, handled by the fourth commit.
- 4. Add so method errors caused by keywords is displayed as the original method, with a text stating if the keyword didn't match. This fixes #15639

Example from #15639:

```
type FooCon
    x
end
type BarCon
    x
end
function addConstraint(c::FooCon)
    println("addCon(FooCon)")
end
function addConstraint(c::BarCon; uncset=nothing)
    println("addCon(BarCon,$uncset)")
    # Oh, BarCon was suppossed to be a FooCon, let me erroneously pass
    # on this kwarg which isn't supported
    addConstraint(FooCon(c.x), uncset=uncset)
end
julia> addConstraint(BarCon(1))
addCon(BarCon,nothing)
ERROR: MethodError: no method matching addConstraint(::FooCon; uncset=nothing)
Closest candidates are:
  addConstraint(::FooCon) got an unrecognized keyword argument "uncset"
  addConstraint(::BarCon; uncset)
 in #addConstraint#1(::Void, ::Function, ::BarCon) at ./REPL[4]:5
 in addConstraint(::BarCon) at ./REPL[4]:2
 in eval(::Module, ::Any) at ./boot.jl:225
 in macro expansion at ./REPL.jl:92 [inlined]
 in (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:46
```

Note to reviewer please check if the changes in `src/julia-syntax.scm` is correct as this is my first encounter with lisp.

### CODE REVIEW

**Comment by IainNZ** [comment]
@dhoegh can you post the new error message for that case?

**Comment by quinnj** [comment]
A big +1 to progress with this. In the method lists, I wonder if we should display the default kwarg value too? Seems like that's usually the next question once I know there is a keyword argument involved.

Also probably want to track the updates in #16580 with how keyword arguments will be handled internally.

**Comment by dhoegh** [comment]
I have updated the example.

File: base/replutil.jl [hunk 1]
```diff
@@ -123,6 +123,16 @@ function showerror(io::IO, ex::MethodError)
     ft = typeof(f)
     name = ft.name.mt.name
     f_is_function = false
+    kwargs = Dict{Any,Any}()
```
Comment (vtjnash) [inline base/replutil.jl]: i think you meant `Array{Any}(0)` here?

Comment (dhoegh) [inline base/replutil.jl]: Yes, thank you. I have corrected it.

**Comment by dhoegh** [comment]
I have fixed the comment from vtnash. I have taken a brief look at #16580 and the approach used here seems to be adaptable for the PR. Is #16580 slated 0.5 material, because I do think it would be good to get this in 0.5.

**Comment by nalimilan** [comment]
Great!

Though for consistency with positional arguments, the value of keyword arguments shouldn't be printed on `MethodError`. So this:
```julia
ERROR: MethodError: no method matching addConstraint(::FooCon; uncset=nothing)
```
would rather be:
```julia
ERROR: MethodError: no method matching addConstraint(::FooCon; uncset::Void)
```

Also, maybe "unsupported" would be better than "unrecognized".

**Comment by IainNZ** [comment]
@nalimilan why would it be Void? Surely its Any?

**Comment by nalimilan** [comment]
@IainNZ AFAIK the displayed type is that of the passed argument, not the one in the signature (indeed, that argument doesn't even exist for that method, so it doesn't have a type).

**Comment by IainNZ** [comment]
That... doesn't make any sense then, right? Doesn't seem like something to perpuate

File: base/replutil.jl [hunk 2]
```diff
@@ -150,6 +160,13 @@ function showerror(io::IO, ex::MethodError)
             print(io, "::$typ")
             i == length(arg_types_param) || print(io, ", ")
         end
+        if 0 < length(kwargs)
```
Comment (tkelman) [inline base/replutil.jl]: use isempty instead of comparing length to 0

**Comment by dhoegh** [comment]
I have addressed the comment from tkelman and changed "unrecognized" to "unsupported" based on nalimilan comment.

Regarding display of default values, that seem to require more than a trivial amount of look through the AST of the `kwsorter` function. Therefore I do not think it is appropriate to implement it here, but it would be a nice addition.

**Comment by dhoegh** [comment]
Bump, any further comments or thoughts.

File: base/replutil.jl [hunk 3]
```diff
@@ -150,6 +160,13 @@ function showerror(io::IO, ex::MethodError)
             print(io, "::$typ")
             i == length(arg_types_param) || print(io, ", ")
         end
+        if !isempty(kwargs)
+            print(io, "; ")
+            for (i ,(k, v)) in enumerate(kwargs)
+                print(io, k, "=", v)
```
Comment (vtjnash) [inline base/replutil.jl]: use `show(IOContext(io, :limit => true), v)`, not `print(io, v)` for `v`

(also the `\space,` are in the wrong order on the previous line)

File: base/replutil.jl [hunk 4]
```diff
@@ -123,6 +123,16 @@ function showerror(io::IO, ex::MethodError)
...
     name = ft.name.mt.name
     f_is_function = false
+    kwargs = Any[]
+    if startswith(string(ft.name), "#kw#")
+        f = ex.args[2]
+        ft = typeof(f)
+        name = ft.name.mt.name
+        arg_types_param = arg_types_param[3:end]
+        temp = ex.args[1]
+        kwargs = [(temp[i*2-1], temp[i*2]) for i in 1:(length(temp) ÷ 2)]
```
Comment (vtjnash) [inline base/replutil.jl]: probably best to declare this as a typed comprehension with output type `Any`

File: base/replutil.jl [hunk 5]
```diff
@@ -222,7 +239,7 @@ function showerror_nostdio(err, msg::AbstractString)
     ccall(:jl_printf, Cint, (Ptr{Void},Cstring), stderr_stream, "\n")
 end
 
-function show_method_candidates(io::IO, ex::MethodError)
+function show_method_candidates(io::IO, ex::MethodError, kwargs=Tuple{Symbol, Any}[])
```
Comment (vtjnash) [inline base/replutil.jl]: I think `kwargs::Vector{Any}` is probably better here

**Comment by vtjnash** [comment]
sorry this took so long to review. I've been coming back to this every few days trying to make sure I was understanding how everything would interact.

Can I ask you to add a test for the behavior with multiple kwargs (multiple matching, multiple not matching)?

**Comment by StefanKarpinski** [comment]
Given that this is a *display* change, which tends to be less likely to break people's code, can't we merge this and then patch it up?

**Comment by dhoegh** [comment]
I have fixed the trivial comments and added test for multiple kwargs matching and multiple not matching at line https://github.com/dhoegh/julia/blob/d4e8792075735c9a71e52f3ddad62cb96bf589ad/test/replutil.jl#L112-L115. I have squashed the commit where the review comments was addressed away. My time is very limited right now, so my response time the next couple of weeks could be long.

**Comment by vtjnash** [comment]
Thanks. This is great.

**Pull request merged**